arlas-commons/src/main/java/io/arlas/filter/impl/AbstractPolicyEnforcer.java

```diff
@@ -109,71 +133,111 @@ private void addTechnicalRolesToPermissions(Set<String> permissions, Map<String,
-                ArlasClaims arlasClaims = new ArlasClaims(permissions.stream().toList());
-                ctx.setProperty("claims", arlasClaims.getRules());
-                if ((ok != null && ok) || arlasClaims.isAllowed(ctx.getMethod(), ctx.getUriInfo().getPath())) {
-                    arlasClaims.injectHeaders(ctx.getHeaders(), transaction);
+                ctx.getHeaders().remove(authConf.headerGroup); // remove it in case it's been set manually
+                Map<String, Object> roles = getRolesClaim(token);
+                Set<String> groups = Collections.emptySet();
+                if (!roles.isEmpty()) {
+                    groups = roles.values().stream()
+                            .map(v -> (List<String>) v)
+                            .flatMap(Collection::stream)
+                            .filter(r -> r.toLowerCase().startsWith("group"))
+                            .collect(Collectors.toSet());
+                    ctx.setProperty("groups", groups.stream().toList());
+                    ctx.getHeaders().put(authConf.headerGroup, groups.stream().toList());
+                    LOGGER.debug("Add Header [" + authConf.headerGroup + ": " + groups + "]");
+                }
+                log = StringUtil.concat(log, String.format(" (groups=%s)", groups));
+
+
+                Set<String> permissions = getPermissionsClaim(token);
+                addTechnicalRolesToPermissions(permissions, roles);
+                log = StringUtil.concat(log, String.format(" (permissions=%s)", permissions));
+                if (!permissions.isEmpty()) {
+                    ArlasClaims arlasClaims = new ArlasClaims(permissions.stream().toList());
+                    ctx.setProperty("claims", arlasClaims.getRules());
+                    if ((ok != null && ok) || arlasClaims.isAllowed(ctx.getMethod(), ctx.getUriInfo().getPath())) {
+                        arlasClaims.injectHeaders(ctx.getHeaders(), transaction);
+                        cacheManager.putDecision(getDecisionCacheKey(ctx, accessToken), Boolean.TRUE);
+                        logUAM(ALLOWED,"granted: " + log);
+                        return;
+                    }
+                }
+                if (isPublic) {
                     cacheManager.putDecision(getDecisionCacheKey(ctx, accessToken), Boolean.TRUE);
+                    logUAM(ALLOWED,"public (with token): " + log);
                     return;
                 }
-            }
-            if (isPublic) {
-                cacheManager.putDecision(getDecisionCacheKey(ctx, accessToken), Boolean.TRUE);
+            } catch (Exception e) {
+                LOGGER.warn("JWT verification failed.", e);
+                if (!isPublic) {
+                    logUAM(DENIED,"unauthorized (invalid token): " + log);
+                    ctx.abortWith(Response.status(UNAUTHORIZED).build());
+                } else {
+                    logUAM(ALLOWED,"public (invalid token): " + log);
```

Comment:
shouldn't be `DENIED` too ?

Reply:
It is public therefore allowed (but we log that the token was invalid).

Reply:
ok LGTM